### PR TITLE
Migrate from rustc-serialize to serde 1.0

### LINKED ISSUE #128: use serde
I should start using `serde` for the automatic serialization component of this crate.

I don't think this means removing support for `rustc-serialize` just yet though.

### PR DESCRIPTION
**Pull request by ordian**
This is a breaking change.
* decode() is renamed to deserialize()
  for consistency with serde
* docopt_macros is updated to latest nightly

Fixes #128, #213

### CODE REVIEW

**Comment by ordian** [comment]
Tests on Travis are failing on rust 1.12. 
Any plans to drop support for this old version?

File: .travis.yml [hunk 1]
```diff
@@ -1,6 +1,5 @@
 language: rust
 rust:
-  - 1.12.0
```
Comment (BurntSushi) [inline .travis.yml]: I understand why we can no longer use Rust 1.12 (`serde_derive` requires Rust 1.15 IIRC), but there's a reason why a specific version is tested in Travis: to ensure we don't *accidentally* bump the minimum Rust version required to build docopt. For this reason, please put the minimum version back. I think Rust 1.15 would be suitable.

File: Cargo.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "docopt"
-version = "0.7.0"  #:version
+version = "0.8.0"  #:version
```
Comment (BurntSushi) [inline Cargo.toml]: Please don't bump the version. I'll do that in a separate commit. :-)

File: docopt_macros/Cargo.toml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "docopt_macros"
-version = "0.7.0"  #:version
+version = "0.8.0"  #:version
```
Comment (BurntSushi) [inline docopt_macros/Cargo.toml]: Same as `docopt`'s Cargo.toml. I'll bump this version in a separate commit.

**Comment by ordian** [comment]
Fixed.
Bumped minimum rust version to 1.15, although I don't know why anyone would use stable, but not latest.
Also, replaced try! with ? (IIRC stable since 1.13).
Thanks!

**Comment by BurntSushi** [comment]
Thanks!

>  although I don't know why anyone would use stable, but not latest.

Because not everyone can upgrade everything all the time. Think about Linux distros, company policies, etc etc.

**Comment by ordian** [comment]
Should I revert version bump in readme as well?

**Comment by BurntSushi** [comment]
@ordian Yes please, good catch.

**Comment by ordian** [comment]
How about enum with struct variant support? E.g.
```rust
#[derive(Debug, Deserialize)]
enum CargoCommand {
    Build {
        flag_debug: bool,
        flag_examples: bool,  
    },
    Run {
        ...
    },
    Test,
    Bench,
    ...
}
```

**Comment by BurntSushi** [comment]
@ordian I'd prefer that be separate. It has been discussed quite a bit on the issue tracker, but all the details have fallen out of my head.

**Comment by BurntSushi** [comment]
Thanks! This is awesome work. :-)

**Comment by nixpulvis** [comment]
:DDDD

**Pull request merged**